### ISSUE TITLE
Bug: @typescript-eslint/no-redeclare Allow types and variables with same name

### ISSUE THREAD

**Issue by joshuambg** [typescript-eslint/typescript-eslint#5540]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.4&sourceType=module&code=FAYwNghgzlAED6BBADgSwMJlQUwHYBdYBvYYAX1PwE9ltYUMs9CBeWa2gewDMEHMcBANykQnXFEL8mBWGyRoBzIUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGU4I4pAdgPYC00iAJogMbwCGZ6UhAZs5ADTa54ACALgE8ADiirQAlkL7EU8cfj4B6IqQrU6DDJETRohaJGwBfEEaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
class _ApiClient {

}

type ApiClient = typeof _ApiClient;

const ApiClient = _ApiClient;
```

### ESLint Config

```javascript
{
  "rules": {

        "no-redeclare": "off",
        "@typescript-eslint/no-redeclare": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error to be reported. TypeScript allows types and variables to share the same name.

### Actual Result

`@typescript-eslint/no-redeclare` rule incorrectly reports an error.

`error  'ApiClient' is already defined  @typescript-eslint/no-redeclare`

### Additional Info

The rule needs to be made more intelligent and follow proper TypeScript paradigms. 

### Versions

| package                                | version |
| -------------------------------------- | ------- |
|     `@typescript-eslint/eslint-plugin`     | `5.35.1`|
|     `@typescript-eslint/eslint-plugin-tslint`     | `5.35.1`|
 |    `@typescript-eslint/parser`     | `5.35.1`|
| `TypeScript`                           | `4.4.4` |
| `ESLint`                               | `7.5.0` |
| `node`                                 | `14.17.3` |

**Comment by bradzacher** [comment]
Thanks for the issue!

We recommend not using this rule because it duplicates an error provided by TS that TS just does better.

Additionally declaring a type and value with the same name is **generally** a code smell as it can create weird or confusing code.

For these two reasons we won't be looking to add new options to the rule.

**Comment by joshuambg** [comment]
@bradzacher I now see this issue is actually reported on the rule page. Please update the rule page with your official recommendation not to use the rule, if that is the official stance.

https://typescript-eslint.io/rules/no-redeclare/

**Comment by bradzacher** [comment]
It is on the rule page, though we don't explicitly spell it out.

If we recommended people use the rule then we would include the rule in our recommended config. But as per the attributes section it is included in neither of our configs.

![Screenshot_20220825-115948.png](https://user-images.githubusercontent.com/7462525/186560511-494d6579-3960-4a90-8838-f022afd73169.png)
